spec/generator/gen-named-callout.spec.js

```diff
@@ -0,0 +1,107 @@
+import {toOnigurumaAst} from '../../dist/index.js';
+import {generate} from '../../dist/generator/generate.js';
+
+describe('generate: Named Callout', () => {
+  function gen(pattern) {
+    return generate(toOnigurumaAst(pattern)).pattern;
+  }
+
+  it('should support all named callout names', () => {
+    const cases = [
+      ['(*FAIL)'],
+      ['(*MISMATCH)'],
+      ['(*SKIP)'],
+      ['(*ERROR)'],
+      ['(*MAX{1})'],
+      ['(*COUNT)'],
+      ['(*TOTAL_COUNT)'],
+      ['(*CMP{5,>=,4})'],
+      // ['(*foo)'], // Error: Invalid callout name "foo"
+    ];
+    for (const [input, expected] of cases) {
+      expect(gen(input)).toBe(expected ?? input);
+    }
+  });
+
+  it('should support tag', () => {
+    const cases = [
+      ['(*FAIL[Tag])'],
+      ['(*MISMATCH[Tag])'],
+      ['(*SKIP[Tag])'],
+      ['(*ERROR[Tag])'],
+      ['(*MAX[Tag]{1})'],
+      ['(*COUNT[Tag])'],
+      ['(*TOTAL_COUNT[Tag])'],
+      ['(*CMP[Tag]{5,>=,4})'],
+    ];
+    for (const [input, expected] of cases) {
+      expect(gen(input)).toBe(expected ?? input);
+    }
+  });
+
+  it('should support empty args', () => {
+    const cases = [
+      ['(*FAIL{})'],
+      ['(*MISMATCH{,})'],
+      ['(*SKIP{,,})'],
+      ['(*ERROR{,,,})'],
+      ['(*MAX{1,,})'],
+      ['(*COUNT{,,2})'],
+      ['(*TOTAL_COUNT{,,3,,})'],
+      ['(*CMP{,1,,==,,,3,,,,})'],
+      // TODO: empty args can be optimized out
```

Comment:
Note that I wouldn't worry about optimization for its own sake in the generator, since the optimizer is available separately. Maybe empty args should be removed, but if so it should be because that's the natural or ideal form for the given set of node properties. Not currently familiar enough with the callout arg system to have an opinion.

Reply:
it would be cleaner to remove redundant commas, leading `0`'s and postfix `+`